src/runtime/virtcontainers/stratovirt.go

```diff
@@ -0,0 +1,1306 @@
+		}
+		s.state.mmioBlkSlots[idx] = !isPut
+
+		return idx, nil
+	}
+
+	return 0, fmt.Errorf("failed to setup mmio slot, Name is invalid %q", Name)
+}
+
+func (s *stratovirt) getDevSlot(Name string) (int, error) {
+	slot, err := s.setupMmioSlot(Name, false)
+	if err != nil {
+		return 0, err
+	}
+
+	return slot, nil
+}
+
+func (s *stratovirt) delDevSlot(Name string) error {
+	if _, err := s.setupMmioSlot(Name, true); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *stratovirt) hotplugBlk(ctx context.Context, drive *config.BlockDrive, op Operation) error {
+	err := s.qmpSetup()
+	if err != nil {
+		return err
+	}
+
+	driver := "virtio-blk-mmio"
+
+	defer func() {
+		if err != nil {
+			s.qmpMonitorCh.qmp.ExecuteBlockdevDel(s.qmpMonitorCh.ctx, drive.ID)
+			if errDel := s.delDevSlot(drive.VirtPath); errDel != nil {
+				s.Logger().WithError(errDel).Warn("Failed to delete device slot.")
+			}
+		}
+	}()
+
+	switch op {
+	case AddDevice:
+		sblkDevice := govmmQemu.BlockDevice{
+			ID:       drive.ID,
+			File:     drive.File,
+			ReadOnly: drive.ReadOnly,
+			AIO:      govmmQemu.BlockDeviceAIO("native"),
+		}
+		if err := s.qmpMonitorCh.qmp.ExecuteBlockdevAdd(s.qmpMonitorCh.ctx, &sblkDevice); err != nil {
+			return err
+		}
+
+		slot, err := s.getDevSlot(drive.VirtPath)
+		if err != nil {
+			return err
+		}
+
+		devAddr := fmt.Sprintf("%d", slot)
+		if err := s.qmpMonitorCh.qmp.ExecutePCIDeviceAdd(s.qmpMonitorCh.ctx, drive.ID, drive.ID, driver, devAddr, "", "", 0, false, false); err != nil {
+			return err
+		}
+	case RemoveDevice:
+		if errDel := s.delDevSlot(drive.VirtPath); errDel != nil {
+			s.Logger().WithError(errDel).Warn("Failed to delete device slot.")
+		}
+		if err := s.qmpMonitorCh.qmp.ExecuteDeviceDel(s.qmpMonitorCh.ctx, drive.ID); err != nil {
+			return err
+		}
+
+	default:
+		return fmt.Errorf("operation is not supported %d", op)
+	}
+
+	return nil
+}
+
+func (s *stratovirt) createVirtiofsDaemon(sharedPath string) (VirtiofsDaemon, error) {
+	virtiofsdSocketPath, err := s.virtiofsSocketPath(s.id)
+	if err != nil {
+		return nil, err
+	}
+
+	if s.config.SharedFS == config.VirtioFSNydus {
+		apiSockPath, err := s.nydusdSocketPath(s.id)
+		if err != nil {
+			return nil, err
+		}
+		nd := &nydusd{
+			path:        s.config.VirtioFSDaemon,
+			sockPath:    virtiofsdSocketPath,
+			apiSockPath: apiSockPath,
+			sourcePath:  sharedPath,
+			debug:       s.config.Debug,
+			extraArgs:   s.config.VirtioFSExtraArgs,
+			startFn:     startInShimNS,
+		}
+		nd.setupShareDirFn = nd.setupPassthroughFS
+		return nd, nil
+	}
+
+	// default use virtiofsd
+	return &virtiofsd{
+		path:       s.config.VirtioFSDaemon,
+		sourcePath: sharedPath,
+		socketPath: virtiofsdSocketPath,
+		extraArgs:  s.config.VirtioFSExtraArgs,
+		cache:      s.config.VirtioFSCache,
+	}, nil
+}
+
+func (s *stratovirt) CreateVM(ctx context.Context, id string, network Network, hypervisorConfig *HypervisorConfig) error {
```

Comment:
It might then be cleaner to add it only once it's needed. I'm not insisting though, this seems just a nit, just saying. :-)

Reply:
I can't simply delete it since it's a part of `type Hypervisor` interface. May be we can replace the `network` param with `_` and some comments, I don‘t know if this way is acceptable.

Reply:
Oh alright, I didn't realise that, I'm sorry. Please ignore my comments here and consider this settled.